util/len-caching-mutex/src/lib.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+extern crate parking_lot;
+
+use std::collections::VecDeque;
+use std::ops::{Deref, DerefMut};
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering;
+
+use parking_lot::{Mutex, MutexGuard};
+
+/// Implement to allow a type with a len() method to be used
+/// with [`LenCachingMutex`](struct.LenCachingMutex.html)
+pub trait Len {
+	fn len(&self) -> usize;
+}
+
+impl<T> Len for Vec<T> {
+	fn len(&self) -> usize { self.len() }
+}
+
+impl<T> Len for VecDeque<T> {
+	fn len(&self) -> usize { self.len() }
+}
+
+/// Can be used in place of a `Mutex` where reading `T`'s `len()` without 
+/// needing to lock, is advantageous. 
+/// When the Guard is released, `T`'s `len()` will be cached.
+/// *The cached `len()` may be at most 1 lock behind current state.
+pub struct LenCachingMutex<T> {
+  data: Mutex<T>,
+  len: AtomicUsize,
+}
+
+impl<T: Len> LenCachingMutex<T> {
+	pub fn new(data: T) -> LenCachingMutex<T> {
+		LenCachingMutex {
+			len: AtomicUsize::new(data.len()),
+			data: Mutex::new(data),
+		}
+	}
+
+	/// Load the most recent value returned from your `T`'s `len()`
+	pub fn load_len(&self) -> usize {
+		self.len.load(Ordering::Relaxed)
```

Comment:
I'd prefer using `SeqCst` ordering here (everywhere). I remember we had some discussions early in Parity Ethereum and afair `Relaxed` is `SeqCst` on x86 anyway, and on ARM we might get some heissenbugs because we use `Relaxed`. The performance gain is probably neglegible, so I'd prefer consistency and bug-less code.

Reply:
We don't actually need an `AtomicUsize` here, since we're protecting both the data and the `len` with the given mutex. So technically we could write something like this
```rust
pub struct LenCachingMutex<T> {
  lock: Mutex<()>,
  data: T,
  len: usize,
}
```
But that would involve some amount of unsafe code. 
So IMHO using `Relaxed` here is fine.

Reply:
@ordian the whole point is to have a simple rule of thumb: never use `Relaxed`, cause it has neglegible performance gains and may cause heisenbugs. I'm fine with `AcqRel` ordering in both places.

Reply:
@tomusdrw @ordian  Thanks for the feedback.  I'm definitely happy to use ```SeqCst```, particularly if it helps people reason about the code.